### PR TITLE
Credential configuration attributes

### PR DESCRIPTION
**Pull request by alexsomesan**
### Description

This change introduces the ability to configure credentials for accessing the Kubernetes cluster from various sources. It aligns closely to the way the classic Kubernetes provider is configured, but introduces a few changes to address pain-points repeatedly reported by users in the past.  

Most notably, it no longer automatically loads configuration from kubeconfig files present in well known locations such as `$HOME/.kube/config` or the `$KUBECONFIG` environment variable. Users can still load configuration from these files by specifically setting their path under the `config_path` attribute.

### Acceptance tests
- [x] Have you added an acceptance test for the functionality being added?
- [x] Have you run the acceptance tests on this branch? (If so, please include the test log in a gist)

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### CODE REVIEW

File: docs/usage.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
+## Provider configuration
+
+The provider accepts the following configuration attributes under the `provider` block.
```
Comment (Redeux) [inline docs/usage.md:3]: I like the type description, but I think we should still denote whether something is required or optional.  Aligned with https://www.terraform.io/docs/providers/kubernetes/index.html#argument-reference

File: docs/usage.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  * `api_version` - (string) Version of the "client.authentication.k8s.io" API which the plugin implements.
+  * `command` - (string) The plugin executable (absolute path, or expects the plugin to be in OS PATH).
+  * `env` - (map string to string) Environment values to set on the plugin process.
+  * `args` - (list of strings) Command line arguments to the plugin command.
+
+Due to the internal design of this provider, access to a responsive API server is required both during PLAN and APPLY. The provider makes calls to the Kubernetes API to retrieve metadata and type information during all stages of Terraform operations.
+
+### Credentials
+
+For authentication, the provider can be configured with identity credentials sourced from either a `kubeconfig` file, explicit values in the `provider` block or a combination of both.
```
Comment (Redeux) [inline docs/usage.md:28]:
```suggestion
For authentication, the provider can be configured with identity credentials sourced from either a `kubeconfig` file, explicit values in the `provider` block, or a combination of both.
```

Added an Oxford comma, turns out I do care about them, despite what Vampire Weekend may say.

File: docs/usage.md (hunk 1 above)
Comment (alexsomesan) [reply docs/usage.md:3]: Good point. Everything is optional so I'll just add that as a note before the list. I feel like adding the word "optional" to all the entries just clutters it up? WDYT?

File: docs/usage.md [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+Setting the `config_path` attribute to the path of a `kubeconfig` file will drive the provider to load it and use the credential values in it. When `config_path` is not set **NO EXTERNAL KUBECONFIG WILL BE LOADED**. Specifically, $KUBECONFIG environment variable is **NOT** considered.
+
+Take note of the `current-context` configured in the file. You can override it using the `config_context` provider attribute.
+
+When combined, the values loaded from the `kubeconfig` file are overridden by the corresponding values configured on `provider` block attributes.
```
Comment (Redeux) [inline docs/usage.md:34]:
```suggestion
If both `kubeconfig` and static credentials are defined in the `provider` block, the provider will prefer the static credentials and ignore the `kubeconfig`.
```

File: docs/usage.md [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  * `env` - (map string to string) Environment values to set on the plugin process.
+  * `args` - (list of strings) Command line arguments to the plugin command.
+
+Due to the internal design of this provider, access to a responsive API server is required both during PLAN and APPLY. The provider makes calls to the Kubernetes API to retrieve metadata and type information during all stages of Terraform operations.
+
+### Credentials
+
+For authentication, the provider can be configured with identity credentials sourced from either a `kubeconfig` file, explicit values in the `provider` block or a combination of both.
+
+Setting the `config_path` attribute to the path of a `kubeconfig` file will drive the provider to load it and use the credential values in it. When `config_path` is not set **NO EXTERNAL KUBECONFIG WILL BE LOADED**. Specifically, $KUBECONFIG environment variable is **NOT** considered.
```
Comment (Redeux) [inline docs/usage.md:30]:
```suggestion
If the `config_path` attribute is set to the path of a `kubeconfig` file, the provider will load it and use the credential values in it. When `config_path` is not set **NO EXTERNAL KUBECONFIG WILL BE LOADED**. Specifically, $KUBECONFIG environment variable is **NOT** considered.
```

File: docs/usage.md [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+Take note of the `current-context` configured in the file. You can override it using the `config_context` provider attribute.
+
+When combined, the values loaded from the `kubeconfig` file are overridden by the corresponding values configured on `provider` block attributes.
+
+Generaly, there are four major options for providing identity information to the provider for authentication purposes:
```
Comment (Redeux) [inline docs/usage.md:36]:
```suggestion 
There are five options for providing identity information to the provider for authentication purposes:
```

File: docs/usage.md [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
```
Comment (Redeux) [inline docs/usage.md:37]:
```suggestion

* a kubeconfig
```

File: examples/credentials-exec-auth/exec-cli.tf [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+    command = "aws" # this is the actual 'aws' cli tool
+
+    args = ["--region", "eu-central-1", "eks", "get-token", "--cluster-name", "my-eks-cluster"]
+
+    env = {
+      # Credentials for the AWS cli tool
+      "AWS_PROFILE"           = "hashicorp"
+      # "AWS_ACCESS_KEY_ID"     = "my-access-key-id"
+      # "AWS_SECRET_ACCESS_KEY" = "my-secret-key-data"
```
Comment (Redeux) [inline examples/credentials-exec-auth/exec-cli.tf:25]: Should these be commented out?

Comment (alexsomesan) [reply examples/credentials-exec-auth/exec-cli.tf:25]: Yes, they are mutually exclusive with `AWS_PROFILE`. It's either one or the other two.

File: docs/usage.md (hunk 1 above)
Comment (alexsomesan) [reply docs/usage.md:3]: Added a line noting everything is optional but a meaningful set of options should be configured.

File: examples/credentials-token/token-cred.tf [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
...FZbVZEUVRDQ0FTSXdEUVlKS29aSWh2Y05BUUVCQlFBRGdnRVBBRENDQVFvQ2dnRUJBT0crCm5qem5EWGNwQi9GMUtreWE0K2pKY1VKRWU2UlpWVys1WFo2OS9IQmRqSFVUdkNZbGlJQXovYU1DV3ByUjBlUmgKR0tPQStoOU5lQ0hwUWNBTWhtcy9ZRlBNeXE2bCtPMS9jcjNHM3lvT3NJQnFtOFBQZnNpZ1NNU1czTmNMMlk0egpsYWJDbW1TUDhLclpZS1hmSjE1djVvLzBMYS9wSDd0Wm5GbVFnOENVK09ybWxUYWdjaXFBbnVPOElDMTBSYXgwClZNcUk5eWE0U2NKaHoxdXZoRlhTc1Q4NFovaHIyNXVTMFRVQjV3ejJvaENlTGZMeW5ocWY3aTF4d0t2dkVIRzEKQWQ4ek9weW1vV2ptREhYek1GbmpQV2pFSm00Wm9qTjhYc2dqTDVrSVE0cm1ab0srQjV6QW1XL1JWd3I5dm5FeQp2K01mZXpCeEk4WFJqSncxbHljQ0F3RUFBYU5DTUVBd0RnWURWUjBQQVFIL0JBUURBZ0trTUIwR0ExVWRKUVFXCk1CUUdDQ3NHQVFVRkJ3TUNCZ2dyQmdFRkJRY0RBVEFQQmdOVkhSTUJBZjhFQlRBREFRSC9NQTBHQ1NxR1NJYjMKRFFFQkN3VUFBNElCQVFBcmxLcERxMjhkWUpNQ2I3cFEvZktjd1FUR0o2MEdiUVpUa3k3U1FPSGpWRkFHczNVVAo3RnRCcVl4bk1CVFY3TXhXS01IaG1rR0YyZVBlbVNjVTFVOVkwR2hPVWJVVmxRVHhDUklVeSt5cjZkQVdYSEk0CjQyYnBQVVU4UDIvUlRDWWV5ZzNKWG1KaWkrMDRZajVmRHZMRnFRSHhpbnpxNDFoYlk2SlpVUnR0OXRmMDNkNWUKd25FNGhEOHdmVzE5ZW5TTGFERE01N2RpWWV2SWI3ZzdHcTc4RnBldTlYWWNiM3VJK3g5bVhFWDYxZDA3S1ZSZgpCYlZaa2pZemk0enI0eHZXbUZQTmMyekorZXJwVTMweDdiLzk4eGt6TkRJcTNNekVJeGQyWUNPSnMrZ1kwNk5aCnNGa2xtNStMTTU3NTU1dWVVc2FnMnBrWjdBTXlKQjFmRXFHdwotLS0tLUVORCBDRVJUSUZJQ0FURS0tLS0tCg==")
```
Comment (Redeux) [inline examples/credentials-token/token-cred.tf:9]: Is this a practice we want to promote?

Comment (alexsomesan) [reply examples/credentials-token/token-cred.tf]: flipped it to loading from file.

File: provider/server.go [hunk 9]
```diff
@@ -105,35 +224,197 @@ func (s *RawProviderServer) UpgradeResourceState(ctx context.Context, req *tfplu
...
+	overrides := &clientcmd.ConfigOverrides{}
+	loader := &clientcmd.ClientConfigLoadingRules{}
 
-	// if no config specified, try the known default locations
-	if configFile.IsNull() || configFile.AsString() == "" {
-		h := os.Getenv("HOME")
-		if h == "" {
-			h = os.Getenv("USERPROFILE") // windows
+	var configPath cty.Value
+	if configPathEnv, ok := os.LookupEnv("KUBE_CONFIG_PATH"); ok && configPathEnv != "" {
```
Comment (jrhouston) [inline provider/server.go:236]: I'm still a bit confused as to why we're not supporting `KUBECONFIG` but we are reimplementing it as `KUBE_CONFIG_PATH`. Like what's the use-case you foresee for this? Given that all these `KUBE_` env vars are not standard to kubectl couldn't the user just supply variables to the provider block and use `TF_` env vars if they want to be able to configure the provider through the environment?

File: provider/provider_config.go [hunk 10]
```diff
@@ -3,38 +3,59 @@ package provider
...
+			"args":        cty.List(cty.String),
+		}),
 	})
 }
 
 // GetProviderConfigSchema contains the definitions of all configuration attributes
 func GetProviderConfigSchema() *tfplugin5.Schema {
-	cfgFileType, _ := cty.String.MarshalJSON()
-	boolType, _ := cty.Bool.MarshalJSON()
+	b, _ := ctyObjectToTfpluginSchema(getConfigObjectType())
```
Comment (jrhouston) [inline provider/provider_config.go:36]: Should we be handling this error here? It seems unlikely that the cty type is going to fail to marshal to JSON so maybe we should just have `ctyObjectToTfpluginSchema` only return the Schema_Block and explicitly silence the error there.

File: provider/server.go (hunk 9 above)
Comment (alexsomesan) [reply provider/server.go:236]: KUBECONFIG has a high chance of being inadvertently picked up and loaded "under the radar" and issue history has shown this catches some users on the wrong foot. I decided to experiment with being completely explicit and requiring users to deliberately opt into loading it. I'm looking to see if this is a better UX in the end. We can always reverse to being implicit if there is demand for it.

File: provider/provider_config.go (hunk 10 above)
Comment (alexsomesan) [reply provider/provider_config.go:36]: Good catch. This is a forgotten quick-hack. I'll fix it.

Comment (alexsomesan) [reply provider/provider_config.go]: I added error propagation all the way to the plugin API. It should stop Terraform and surface the error if it ever happens.

**Pull request merged**